Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,98 @@
+//
+// This source file is part of the Swift Metrics API open source project
+//
+// Copyright (c) 2018-2019 Apple Inc. and the Swift Metrics API project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of Swift Metrics API project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+
+public enum SystemMetricsProvider {
+    fileprivate static let lock = ReadWriteLock()
+    fileprivate static let queue = DispatchQueue(label: "com.apple.CoreMetrics.SystemMetricsProvider")
+    fileprivate static var shouldRunSystemMetrics: Bool = false
+
+    public static func bootstrapSystemMetrics() {
+        self.lock.withWriterLockVoid {
+            self.shouldRunSystemMetrics = true
+        }
+        DispatchQueue.global(qos: .background).async {
+            updateSystemMetrics()
+        }
+    }
+    
+    public static func cancelSystemMetrics() {
+        self.lock.withWriterLockVoid {
+            self.shouldRunSystemMetrics = false
+        }
+    }
+    
+    private enum SystemMetricsError: Error {
+        case FileNotFound
+    }
+    
+    fileprivate static func updateSystemMetrics() {
+        self.queue.asyncAfter(deadline: .now() + .seconds(2)) {
+            let shouldReturn = self.lock.withReaderLock { () -> Bool in
+                if !self.shouldRunSystemMetrics {
+                    return true
+                }
+                return false
+            }
+            if shouldReturn { return }
+            let prefix = "process_"
+            let pid = ProcessInfo.processInfo.processIdentifier
+            let ticks = Int32(_SC_CLK_TCK)
+            #if os(Linux)
+            do {
+                guard let statString =
+                    try String(contentsOfFile: "/proc/\(pid)/stat", encoding: .utf8)
+                        .split(separator: ")")
+                        .last
+                    else { throw SystemMetricsError.FileNotFound }
+                let stats = String(statString)
+                        .split(separator: " ")
+                        .map(String.init)
+                if let vmem = Int32(stats[20]) {
+                    Gauge(label: prefix + "virtual_memory_bytes").record(vmem)
+                }
+                if let rss = Int32(stats[21]) {
+                    Gauge(label: prefix + "resident_memory_bytes").record(rss * Int32(_SC_PAGESIZE))
+                }
+                if let startTimeTicks = Int32(stats[19]) {
+                    Gauge(label: prefix + "start_time_seconds").record(startTimeTicks / ticks)
+                }
+                if let utimeTicks = Int32(stats[11]), let stimeTicks = Int32(stats[12]) {
+                    let utime = utimeTicks / ticks, stime = stimeTicks / ticks
+                    Gauge(label: prefix + "cpu_seconds_total").record(utime + stime)
+                }
```

Comment:
I think we should extract the parsing logic to a separate struct with a ctor that takes the raw string and returns well structured "model"

this also accessed array elements unsafely which could lead to index out of bound crashes, so need to guard against that

Reply:
Having two structs - `LinuxOS` and `UnsupportedOS` - would address your suggestion re: "discrete functions" above pretty nicely.